NeoML/src/Dnn/Layers/DnnHeadAdapterLayer.cpp

```diff
@@ -0,0 +1,238 @@
+/* Copyright © 2017-2024 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http ://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-------------------------------------------------------------------------------------------------------------- */
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/NeoML.h>
+#include <NeoML/Dnn/Dnn.h>
+#include <NeoML/Dnn/Layers/DnnHeadAdapterLayer.h>
+#include <NeoML/NeoMLDefs.h>
+
+namespace NeoML {
+
+static const int HeadAdapterLayerVersion = 2000;
+
+void CDnnHeadAdapterLayer::Serialize( CArchive& archive )
+{
+	archive.SerializeVersion(HeadAdapterLayerVersion, CDnn::ArchiveMinSupportedVersion);
+	CBaseLayer::Serialize(archive);
+
+	if(archive.IsStoring()) {
+		archive << num;
+		if(num > 0) {
+			CString name(head->connections[0]->GetName());
+			archive << name;
+			return;
+		}
+
+		internalDnn->Serialize(archive);
+	}
+	else if(archive.IsLoading()) {
+		archive >> num;
+		if(num > 0) {
+			archive >> firstAdapter;
+			return;
+		}
+
+		CDnn* internalDnn = new CDnn(CRandom(0), MathEngine());
+		internalDnn->Serialize(archive);
+
+		head = new CHeadStruct(internalDnn);
+		SetDnnHead(head);
+	}
+	else {
+		NeoAssert(false);
+	}
+}
+
+void CDnnHeadAdapterLayer::OnDnnChanged(CDnn*)
+{
+	if(head == nullptr) {
+		if(!firstAdapter.IsEmpty()) {
+			SetDnnHead(static_cast<CDnnHeadAdapterLayer*>((GetDnn()->GetLayer(firstAdapter).Ptr()))->head);
+		}
+	}
+}
+
+void CDnnHeadAdapterLayer::Reshape()
+{
+	if(num > 0) {
+		if(head->connections[0]->IsLearningEnabled()) {
+			EnableLearning();
+		} else {
+			DisableLearning();
+		}
+		outputDescs[0] = internalDnn->GetLayer("sink")->inputDescs[0];
+		return;
+	}
+
+	auto headSource = static_cast<CCompositeSourceLayer*>(internalDnn->GetLayer("source").Ptr());
+	if(headSource->GetBackwardForced() != IsBackwardNeeded()) {
+		headSource->SetBackwardForced(IsBackwardNeeded());
+	}
+	headSource->SetBlobDesc(inputDescs[0]);
+
+	setInternalDnnParams();
+	calcBlobsForBackwardAndLearn();
+	outputDescs[0] = internalDnn->GetLayer("sink")->inputDescs[0];
+}
+
+void CDnnHeadAdapterLayer::RunOnce()
+{
+	innerInputBlobs.DeleteAll();
+	NeoAssert(inputBlobs.Size() == 1);
+
+	auto headSource = static_cast<CCompositeSourceLayer*>(internalDnn->GetLayer("source").Ptr());
+	headSource->SetBlob(inputBlobs[0]);
+	internalDnn->runOnce( 0 );
```

Comment:
The special test should be where the `CDnnHeadAdapterLayer` is inside the `CRecursiveLayer`.